pkg/controller/plugins.go

```diff
@@ -0,0 +1,228 @@
+	// Name is the name of the plugin
+	Name string
+	// Running is the runtime status of the plugin.  See managedPlugin.running
+	Running bool
+	// Err is any error caught for this plugin.  See managedPlugin.err
+	Err error
+}
+
+// ControllerConfig is how controller plugins receive configuration.  An instance of ControllerConfig specific to
+// the plugin will be passed to the plugin's ProbeControllerPlugins(config) func.  Only the binary hosting the
+// plugins knows what plugins are included.  This requires specific configuration to happen in the binary which
+// is passed downward generically via ControllerConfig.
+//
+// Volumes in ControllerConfig are intended to be relevant to several plugins, but not necessarily all plugins.
+// The preference is to leverage strong typing in this struct.  All config items must have a descriptive but
+// non-specific name (e.g, SyncPeriod is OK but SyncPeriodForNodeController is !OK).
+//
+// OtherAttributes is a map of string values intended for one-off configuration relevant to a single plugin. All
+// values are passed by string and require interpretation by the plugin.  Passing config strings is the least
+// desirable option.  The preference is for common attributes and strong typing for config.
+type ControllerConfig struct {
+	SyncPeriod      time.Duration
+	OtherAttributes map[string]string
+}
+
+// InitPlugins initializes each plugin.
+// This must be called exactly once before any New* methods are called on any plugins.
+func (mgr *pluginMgr) InitPlugins(plugins []Plugin, host Host) error {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+
+	allErrs := []error{}
+	for _, plugin := range plugins {
+		name := plugin.Name()
+		if !util.IsQualifiedName(name) {
+			allErrs = append(allErrs, fmt.Errorf("controller plugin has invalid name: %#v", plugin))
+			continue
+		}
+
+		if _, found := mgr.plugins[name]; found {
+			allErrs = append(allErrs, fmt.Errorf("controller plugin %q was registered more than once", name))
+			continue
+		}
+
+		plugin.Init(host)
+		mgr.plugins[name] = &managedPlugin{plugin: plugin}
+		glog.V(1).Infof("Loaded controller plugin %q", name)
+	}
+	return errors.NewAggregate(allErrs)
+}
+
+// FindPluginByName fetches a plugin by name.  If no plugin is found, returns error.
+func (mgr *pluginMgr) FindPluginByName(name string) (Plugin, error) {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+	if managedPlugin, ok := mgr.plugins[name]; ok {
+		return managedPlugin.plugin, nil
+	}
+	return nil, fmt.Errorf("no controller plugin found matching %s", name)
+}
+
+// RunAll attempts to create Controller instances for each plugin and run them.  This method will not fail
+// for any particular plugin.  Runtime Status is kept for all plugins.  Callers should use pluginMgr.Status
+func (mgr *pluginMgr) RunAll() {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+	for name, managedPlugin := range mgr.plugins {
+		// do not attempt to run a plugin that had previous errors or is already running.
```

Comment:
So if one plugin has an error the only way to try again is to restart the whole binary?